services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedRegistrationService.java

```diff
@@ -472,33 +473,41 @@ private RegistrationResult getRegistrationResult(final String deviceId, final Js
                                 findDevice(tenantId, deviceId))));
     }
 
-    private <T> Future<T> isMaxDevicesLimitReached(final String tenantId) {
+    private Future<Void> checkDeviceLimitReached(final Tenant tenant, final String tenantId) {
 
-        if (config.getMaxDevicesPerTenant() == MongoDbBasedRegistrationConfigProperties.UNLIMITED_DEVICES_PER_TENANT) {
+        final boolean isLimitedAtTenantLevel = Optional.ofNullable(tenant.getRegistrationLimits())
+                .map(RegistrationLimits::isNumberOfDevicesLimited)
+                .orElse(false);
+
+        if (!config.isNumberOfDevicesPerTenantLimited() && !isLimitedAtTenantLevel) {
             return Future.succeededFuture();
         }
 
+        final int maxNumberOfDevices;
+        if (isLimitedAtTenantLevel) {
+            maxNumberOfDevices = tenant.getRegistrationLimits().getMaxNumberOfDevices();
+        } else {
+            maxNumberOfDevices = config.getMaxDevicesPerTenant();
```

Comment:
While checking for maximum number of devices, if tenant level limit is not configured then device registry level configuration is used. In case of maximum number of credentials, provision to configure at the registry level is missing. IMHO it is better to have such a provision to limit the credentials too. Right now if tenant level is not configured then unlimited number of credentials can be added. If we have to enforce a default limit other than unlimited then we have to configure each and every tenants. Having a provision at the device registry level for the credentials limit would easen up things.

Reply:
> IMHO it is better to have such a provision to limit the credentials too.

We currently do not have such a feature and nobody has asked for it (yet). Feel free to create another PR for adding this ...